src/codeu/chat/util/PasswordStorage.java

```diff
@@ -0,0 +1,241 @@
+        byte[] testHash = pbkdf2(password, salt, iterations, hash.length);
+        // Compare the hashes in constant time. The password is correct if
+        // both hashes match.
+        return slowEquals(hash, testHash);
+    }
+
+    private static boolean slowEquals(byte[] a, byte[] b)
+    {
+        int diff = a.length ^ b.length;
+        for(int i = 0; i < a.length && i < b.length; i++)
+            diff |= a[i] ^ b[i];
+        return diff == 0;
+    }
+
+    private static byte[] pbkdf2(char[] password, byte[] salt, int iterations, int bytes)
+        throws CannotPerformOperationException
+    {
+        try {
+            PBEKeySpec spec = new PBEKeySpec(password, salt, iterations, bytes * 8);
+            SecretKeyFactory skf = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
+            return skf.generateSecret(spec).getEncoded();
+        } catch (NoSuchAlgorithmException ex) {
+            throw new CannotPerformOperationException(
+                "Hash algorithm not supported.", ex);
+        } catch (InvalidKeySpecException ex) {
+            throw new CannotPerformOperationException(
+                "Invalid key spec.", ex);
+        }
+    }
+
+    private static byte[] fromBase64(String hex)
+        throws IllegalArgumentException
+    {
+        return DatatypeConverter.parseBase64Binary(hex);
+    }
+
+    private static String toBase64(byte[] array)
+    {
+        return DatatypeConverter.printBase64Binary(array);
+    }
+
+    /* My added methods */
+
+    public static String setPassword() {
+        Console console = System.console();
+        char password[] = console.readPassword("Enter a new password: ");
+        char passwordConfirm[] = console.readPassword("Retype your password: ");
+        System.out.println("Verifying...");
+        while (!Arrays.equals(password, passwordConfirm)) {
+            System.out.println("Passwords didn't match. Please try again.");
+            password = console.readPassword("Enter a new password: ");
+            passwordConfirm = console.readPassword("Retype your password: ");
+            System.out.println("Verifying...");
+        }
+        try {
+            return PasswordStorage.createHash(password);
+        } catch (PasswordStorage.CannotPerformOperationException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    public static void authenticate(String password) {
+        Console console = System.console();
+        char passwordArray[] = console.readPassword("Enter your password: ");
+        System.out.println("Verifying...");
+        try {
+            while (!PasswordStorage.verifyPassword(passwordArray, password)) {
+                System.out.println("Login failed. Please try again");
+                passwordArray = console.readPassword("Enter your password: ");
+                System.out.println("Verifying...");
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
also, when you throw this exception, the client doesn't know if the password matched or not, and creates a Admin panel anyway, instead of staying in the root panel.